### ISSUE TITLE
3.3.4: pytest and tox are failing

### ISSUE THREAD

**Issue by kloczek** [Python-Markdown/markdown#1182]
I'm trying to package your module as an rpm package. So I'm using the typical build, install and test cycle used on building packages from non-root account.
- "setup.py build"
- "setup.py install --root </install/prefix>"
- "pytest with PYTHONPATH pointing to sitearch and sitelib inside </install/prefix>

On testting usimg pytest it fails with:
```console
+ /usr/bin/pytest -ra tests
=========================================================================== test session starts ============================================================================
platform linux -- Python 3.8.12, pytest-6.2.5, py-1.10.0, pluggy-0.13.1
benchmark: 3.4.1 (defaults: timer=time.perf_counter disable_gc=False min_rounds=5 min_time=0.000005 max_time=1.0 calibration_precision=10 warmup=False warmup_iterations=100000)
Using --randomly-seed=1062042675
rootdir: /home/tkloczko/rpmbuild/BUILD/markdown-3.3.4
plugins: forked-1.3.0, shutil-1.7.0, virtualenv-1.7.0, expect-1.1.0, flake8-1.0.7, timeout-1.4.2, betamax-0.8.1, freezegun-0.4.2, aspectlib-1.5.2, toolbox-0.5, rerunfailures-9.1.1, requests-mock-1.9.3, cov-2.12.1, flaky-3.7.0, benchmark-3.4.1, xdist-2.3.0, pylama-7.7.1, datadir-1.3.1, regressions-2.2.0, xprocess-0.18.1, black-0.3.12, asyncio-0.15.1, subtests-0.5.0, isort-2.0.0, hypothesis-6.14.6, mock-3.6.1, profiling-1.7.0, randomly-3.8.0, nose2pytest-1.0.8, pyfakefs-4.5.1, tornado-0.8.1, twisted-1.13.3, aiohttp-0.3.0, localserver-0.5.0, anyio-3.3.1, trio-0.7.0, cases-3.6.4, yagot-0.5.0, Faker-8.14.0
collected 1043 items / 1 error / 1042 selected

================================================================================== ERRORS ==================================================================================
_____________________________________________________ ERROR collecting tests/test_syntax/extensions/test_md_in_html.py _____________________________________________________
/usr/lib/python3.8/site-packages/_pytest/runner.py:311: in from_call
    result: Optional[TResult] = func()
/usr/lib/python3.8/site-packages/_pytest/runner.py:341: in <lambda>
    call = CallInfo.from_call(lambda: list(collector.collect()), "collect")
/usr/lib/python3.8/site-packages/_pytest/python.py:766: in collect
    self.warn(
/usr/lib/python3.8/site-packages/_pytest/nodes.py:230: in warn
    warnings.warn_explicit(
E   pytest.PytestCollectionWarning: cannot collect test class 'TestSuite' because it has a __init__ constructor (from: tests/test_syntax/extensions/test_md_in_html.py)
========================================================================= short test summary info ==========================================================================
ERROR tests/test_syntax/extensions/test_md_in_html.py::TestSuite - pytest.PytestCollectionWarning: cannot collect test class 'TestSuite' because it has a __init__ constr...
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
============================================================================= 1 error in 0.99s =============================================================================
pytest-xprocess reminder::Be sure to terminate the started process by running 'pytest --xkill' if you have not explicitly done so in your fixture with 'xprocess.getinfo(<process_name>).terminate()'.
```
Runnimg pytest with `--import-mode=importli` it fails with:
```console
+ /usr/bin/pytest -ra --import-mode=importlib tests
=========================================================================== test session starts ============================================================================
platform linux -- Python 3.8.12, pytest-6.2.5, py-1.10.0, pluggy-0.13.1
benchmark: 3.4.1 (defaults: timer=time.perf_counter disable_gc=False min_rounds=5 min_time=0.000005 max_time=1.0 calibration_precision=10 warmup=False warmup_iterations=100000)
Using --randomly-seed=3873927089
rootdir: /home/tkloczko/rpmbuild/BUILD/markdown-3.3.4
plugins: forked-1.3.0, shutil-1.7.0, virtualenv-1.7.0, expect-1.1.0, flake8-1.0.7, timeout-1.4.2, betamax-0.8.1, freezegun-0.4.2, aspectlib-1.5.2, toolbox-0.5, rerunfailures-9.1.1, requests-mock-1.9.3, cov-2.12.1, flaky-3.7.0, benchmark-3.4.1, xdist-2.3.0, pylama-7.7.1, datadir-1.3.1, regressions-2.2.0, xprocess-0.18.1, black-0.3.12, asyncio-0.15.1, subtests-0.5.0, isort-2.0.0, hypothesis-6.14.6, mock-3.6.1, profiling-1.7.0, randomly-3.8.0, nose2pytest-1.0.8, pyfakefs-4.5.1, tornado-0.8.1, twisted-1.13.3, aiohttp-0.3.0, localserver-0.5.0, anyio-3.3.1, trio-0.7.0, cases-3.6.4, yagot-0.5.0, Faker-8.14.0
collected 764 items / 1 error / 763 selected

================================================================================== ERRORS ==================================================================================
_____________________________________________________ ERROR collecting tests/test_syntax/extensions/test_md_in_html.py _____________________________________________________
ImportError while importing test module '/home/tkloczko/rpmbuild/BUILD/markdown-3.3.4/tests/test_syntax/extensions/test_md_in_html.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
tests/test_syntax/extensions/test_md_in_html.py:25: in <module>
    from ..blocks.test_html_blocks import TestHTMLBlocks
E   ImportError: attempted relative import with no known parent package
========================================================================= short test summary info ==========================================================================
ERROR tests/test_syntax/extensions/test_md_in_html.py
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
============================================================================= 1 error in 0.87s =============================================================================
pytest-xprocess reminder::Be sure to terminate the started process by running 'pytest --xkill' if you have not explicitly done so in your fixture with 'xprocess.getinfo(<process_name>).terminate()'.
```
tox is failing as well
```conosle
+ /usr/bin/tox --skip-missing-interpreters
.package create: /home/tkloczko/rpmbuild/BUILD/markdown-3.3.4/.tox/.package
ERROR: invocation failed (exit code 1), logfile: /home/tkloczko/rpmbuild/BUILD/markdown-3.3.4/.tox/.package/log/.package-0.log
================================================================================ log start =================================================================================
RuntimeError: failed to build image wheel because:
Traceback (most recent call last):
  File "/usr/lib/python3.8/site-packages/virtualenv/seed/embed/via_app_data/via_app_data.py", line 51, in _install
    key = Path(installer_class.__name__) / wheel.path.stem
AttributeError: 'NoneType' object has no attribute 'path'


================================================================================= log end ==================================================================================
ERROR: InvocationError for command /usr/bin/python3 -m virtualenv --no-download --python /usr/bin/python3 .package (exited with code 1)
ERROR: FAIL could not package project - v = InvocationError('/usr/bin/python3 -m virtualenv --no-download --python /usr/bin/python3 .package', 1)
```

**Comment by kloczek** [comment]
Additional Q: How to build module documentation as man page?

**Comment by facelessuser** [comment]
You seem to be running the tests with pytest which is not supported. Please read development documentation: https://python-markdown.github.io/contributing/#development-environment.

**Comment by waylan** [comment]
I'm closing this as there has been no follow-up. Note that...

1. As previously stated, we do not support pytest so any reports about it not working will be ignored.
2. We are not able to reproduce the report that tox is not running correctly. Presumably, we would need more information about the environment to replicate the issue.

> Additional Q: How to build module documentation as man page?

I'm not sure if you are referring to the CLI or the `docs/` directory. For the CLI, we are using `optparse` from the Python Standard Library. There should be various existing third-party tools which will convert the output of `optparse`'s `--help` to man page(s). I would suggest exploring those options. We currently use MkDocs to render the content of the `docs/` directory to HTML. Last I knew, no-one has made a MkDocs plugin which outputs man pages and I could care less about doing so.

**Comment by kloczek** [comment]
> You seem to be running the tests with pytest which is not supported

I want only to add that running test suite in fixed env has some value however it is some other value to run test suite in opened set of modules if pytest is uses because power of that framework comes from extension which can alter testing process without single touch tested tree.
I'm not asking to switch CI to pytest however I think that alter test framework to be able tun pytest in some minimal env may have some added value.
Looks like it is issue with only with tests/test_syntax/extensions/test_md_in_html.py.
As long as that is added to ignore list pytest Work™️.
```console
+ PYTHONPATH=/home/tkloczko/rpmbuild/BUILDROOT/python-markdown-3.3.6-2.fc35.x86_64/usr/lib64/python3.8/site-packages:/home/tkloczko/rpmbuild/BUILDROOT/python-markdown-3.3.6-2.fc35.x86_64/usr/lib/python3.8/site-packages
+ /usr/bin/pytest -ra --ignore tests/test_syntax/extensions/test_md_in_html.py
=========================================================================== test session starts ============================================================================
platform linux -- Python 3.8.12, pytest-6.2.5, py-1.11.0, pluggy-1.0.0
rootdir: /home/tkloczko/rpmbuild/BUILD/markdown-3.3.6
plugins: datadir-1.3.1, regressions-2.2.0, mock-3.6.1
collected 772 items

tests/test_apis.py ......................................................................................                                                            [ 11%]
tests/test_extensions.py .....................................                                                                                                       [ 15%]
tests/test_legacy.py ................................................................................sssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssss [ 34%]
sssss..s.......................                                                                                                                                      [ 38%]
tests/test_meta.py ..                                                                                                                                                [ 38%]
tests/test_syntax/blocks/test_blockquotes.py .                                                                                                                       [ 38%]
tests/test_syntax/blocks/test_code_blocks.py .....                                                                                                                   [ 39%]
tests/test_syntax/blocks/test_headers.py ss.........................................................................                                                 [ 49%]
tests/test_syntax/blocks/test_hr.py ...............................................                                                                                  [ 55%]
tests/test_syntax/blocks/test_html_blocks.py ..............................................................................................................          [ 69%]
tests/test_syntax/blocks/test_paragraphs.py .................                                                                                                        [ 71%]
tests/test_syntax/extensions/test_abbr.py .............                                                                                                              [ 73%]
tests/test_syntax/extensions/test_admonition.py ......                                                                                                               [ 74%]
tests/test_syntax/extensions/test_attr_list.py ..                                                                                                                    [ 74%]
tests/test_syntax/extensions/test_code_hilite.py sssssssssssssssssssssssssssssssssss                                                                                 [ 79%]
tests/test_syntax/extensions/test_def_list.py ......                                                                                                                 [ 79%]
tests/test_syntax/extensions/test_fenced_code.py ..................ssssssssssssss                                                                                    [ 83%]
tests/test_syntax/extensions/test_footnotes.py ..........                                                                                                            [ 85%]
tests/test_syntax/extensions/test_legacy_attrs.py .                                                                                                                  [ 85%]
tests/test_syntax/extensions/test_legacy_em.py ......                                                                                                                [ 86%]
tests/test_syntax/extensions/test_tables.py .                                                                                                                        [ 86%]
tests/test_syntax/extensions/test_toc.py .........................                                                                                                   [ 89%]
tests/test_syntax/inline/test_autolinks.py .....                                                                                                                     [ 90%]
tests/test_syntax/inline/test_emphasis.py .................                                                                                                          [ 92%]
tests/test_syntax/inline/test_entities.py ....                                                                                                                       [ 92%]
tests/test_syntax/inline/test_images.py ......................                                                                                                       [ 95%]
tests/test_syntax/inline/test_links.py ................................                                                                                              [ 99%]
tests/test_syntax/inline/test_raw_html.py .                                                                                                                          [100%]

========================================================================= short test summary info ==========================================================================
SKIPPED [19] markdown/test_tools.py:185: Excluded
SKIPPED [50] markdown/test_tools.py:144: Tidylib not available.
SKIPPED [1] tests/test_syntax/blocks/test_headers.py:112: This is broken in Python-Markdown
SKIPPED [1] tests/test_syntax/blocks/test_headers.py:132: This is broken in Python-Markdown
SKIPPED [1] tests/test_syntax/extensions/test_code_hilite.py:116: Pygments== is required
SKIPPED [1] tests/test_syntax/extensions/test_code_hilite.py:55: Pygments== is required
SKIPPED [1] tests/test_syntax/extensions/test_code_hilite.py:69: Pygments== is required
SKIPPED [1] tests/test_syntax/extensions/test_code_hilite.py:85: Pygments== is required
SKIPPED [1] tests/test_syntax/extensions/test_code_hilite.py:150: Pygments== is required
SKIPPED [1] tests/test_syntax/extensions/test_code_hilite.py:141: Pygments== is required
SKIPPED [1] tests/test_syntax/extensions/test_code_hilite.py:174: Pygments== is required
SKIPPED [1] tests/test_syntax/extensions/test_code_hilite.py:283: Pygments== is required
SKIPPED [1] tests/test_syntax/extensions/test_code_hilite.py:215: Pygments== is required
SKIPPED [1] tests/test_syntax/extensions/test_code_hilite.py:318: Pygments== is required
SKIPPED [1] tests/test_syntax/extensions/test_code_hilite.py:301: Pygments== is required
SKIPPED [1] tests/test_syntax/extensions/test_code_hilite.py:187: Pygments== is required
SKIPPED [1] tests/test_syntax/extensions/test_code_hilite.py:200: Pygments== is required
SKIPPED [1] tests/test_syntax/extensions/test_code_hilite.py:159: Pygments== is required
SKIPPED [1] tests/test_syntax/extensions/test_code_hilite.py:269: Pygments== is required
SKIPPED [1] tests/test_syntax/extensions/test_code_hilite.py:228: Pygments== is required
SKIPPED [1] tests/test_syntax/extensions/test_code_hilite.py:256: Pygments== is required
SKIPPED [1] tests/test_syntax/extensions/test_code_hilite.py:243: Pygments== is required
SKIPPED [1] tests/test_syntax/extensions/test_code_hilite.py:99: Pygments== is required
SKIPPED [1] tests/test_syntax/extensions/test_code_hilite.py:335: Pygments== is required
SKIPPED [1] tests/test_syntax/extensions/test_code_hilite.py:134: Pygments== is required
SKIPPED [1] tests/test_syntax/extensions/test_code_hilite.py:359: Pygments== is required
SKIPPED [1] tests/test_syntax/extensions/test_code_hilite.py:556: Pygments== is required
SKIPPED [1] tests/test_syntax/extensions/test_code_hilite.py:581: Pygments== is required
SKIPPED [1] tests/test_syntax/extensions/test_code_hilite.py:604: Pygments== is required
SKIPPED [1] tests/test_syntax/extensions/test_code_hilite.py:479: Pygments== is required
SKIPPED [1] tests/test_syntax/extensions/test_code_hilite.py:543: Pygments== is required
SKIPPED [1] tests/test_syntax/extensions/test_code_hilite.py:530: Pygments== is required
SKIPPED [1] tests/test_syntax/extensions/test_code_hilite.py:398: Pygments== is required
SKIPPED [1] tests/test_syntax/extensions/test_code_hilite.py:418: Pygments== is required
SKIPPED [1] tests/test_syntax/extensions/test_code_hilite.py:459: Pygments== is required
SKIPPED [1] tests/test_syntax/extensions/test_code_hilite.py:435: Pygments== is required
SKIPPED [1] tests/test_syntax/extensions/test_code_hilite.py:377: Pygments== is required
SKIPPED [1] tests/test_syntax/extensions/test_code_hilite.py:630: Pygments== is required
SKIPPED [1] tests/test_syntax/extensions/test_code_hilite.py:517: Pygments== is required
SKIPPED [1] tests/test_syntax/extensions/test_fenced_code.py:537: Pygments== is required
SKIPPED [1] tests/test_syntax/extensions/test_fenced_code.py:414: Pygments== is required
SKIPPED [1] tests/test_syntax/extensions/test_fenced_code.py:573: Pygments== is required
SKIPPED [1] tests/test_syntax/extensions/test_fenced_code.py:450: Pygments== is required
SKIPPED [1] tests/test_syntax/extensions/test_fenced_code.py:609: Pygments== is required
SKIPPED [1] tests/test_syntax/extensions/test_fenced_code.py:485: Pygments== is required
SKIPPED [1] tests/test_syntax/extensions/test_fenced_code.py:683: Pygments== is required
SKIPPED [1] tests/test_syntax/extensions/test_fenced_code.py:735: Pygments== is required
SKIPPED [1] tests/test_syntax/extensions/test_fenced_code.py:708: Pygments== is required
SKIPPED [1] tests/test_syntax/extensions/test_fenced_code.py:760: Pygments== is required
SKIPPED [1] tests/test_syntax/extensions/test_fenced_code.py:506: Pygments== is required
SKIPPED [1] tests/test_syntax/extensions/test_fenced_code.py:665: Pygments== is required
SKIPPED [1] tests/test_syntax/extensions/test_fenced_code.py:644: Pygments== is required
SKIPPED [1] tests/test_syntax/extensions/test_fenced_code.py:384: Pygments== is required
===================================================================== 652 passed, 120 skipped in 3.52s =====================================================================
```
In other words to make whole tree pytest ready it does not look that is that much to do ..

**Comment by facelessuser** [comment]
I can understand the desire to use a specific testing framework, but that choice is chosen by the maintainer and must be respected.

If there was a specific advantage for us to move over to pytest, we could have that conversation, but I realize, as you mentioned, that is not the conversation we are currently having.

If we were to add py.test support just so others could use it, then we have to maintain it and ensure it doesn't break or we are going to get issues to "fix pytest" on occasions when it breaks inadvertently. Since this project is not using any features from pytest, it makes no sense for us to go out of our way, even if it is only slightly out of the way, to accommodate pytest.

I have nothing against pytest and use it in a number of my projects, but this project's maintainer has chosen not to use it. I can see no upside for us to fix pytest support as we have no intention of monitoring ensuring that it does not break.

**Comment by kloczek** [comment]
> If we were to add py.test support just so others could use it, then we have to maintain it and ensure it doesn't break or we are going to get issues to "fix pytest" on occasions when it breaks inadvertently. Since this project is not using any features from pytest, it makes no sense for us to go out of our way, even if it is only slightly out of the way, to accommodate pytest.

Adding correctly working pytest support is VERY important.
That importance is not only about exact tested module. Just please add to testing picture fact that in exact test suite is adding yet another workload to all modules involved in testing. In other words it is about health of whole ecosystem (in this case python ecosystem).
Additionally .. I 'm finishing now some packages build automation to be able to run kind of set f scratch build if build systems are not bus. In case of python it address some random set of additional pytest extension and perform build with logging results .. only. In such cases if for example if lets say will be added `pytest-randomly` it would be possible to prove that random reshuffling units may cause that test suite fail it will mean that internally test suite have some dependencies between units and even if all test suite executed without `randomly` extension running exact unit by `pytest <some/file.py>::<some_unit>` will definitely fail. Sometimes test suites are running quite long time or on implementing new unit such way of running isolated units may be useful.
I want only to say that as long that `markdown` test suite will be not working I would not be able to do that or in this exact case I would be doing that only with part of the test suite.

Form that point of view I'm asking to to fix test suite or something like that.
As I have now `markdown` in set of my distro packages I'll be to perform such new way of testing your module.
In other words I want only to encourage to prepare some base line id you want to have such testing completely for free (that automation will be only flagging some found issue without single second spend by you :) )
If you don't care about that is 100%/perfectly fine :)
If you care .. just please have look closer on the code which you know because cost of that fix may be really low :)
I'm specialist on packaging area so form me fixing that seems is a bit to high :P (I've stumped on python modules only ~10 months ago so I'm still learning however after finishing +700 modules as rpm/Solaris IPS packages I see some patters which I can exploit giving back to the community results of my work).
Cheers :)

**Comment by facelessuser** [comment]
> I want only to say that as long that markdown test suite will be not working I would not be able to do that or in this exact case I would be doing that only with part of the test suite.

I understand you have a vision and you have a desire to collect testing data, but I'm going to be very blunt about this, if your whole system depends on every python package you use to support your specific testing framework, you are building a very fragile system. Cornering other projects into using one specific testing framework is not a good idea, and it takes away the freedom of that project to maintain their project in a way that they find most suitable to them.

What if some other framework comes out that a project finds more enticing than pytest? Are they not allowed to take advantage of that testing framework as to not break your pytest dependent system?

I'm not seeing an advantage for us to switch to pytest, and I definitely can't see a reason to set up and test an additional framework that will test the exact same tests.

Again, I have nothing against pytest, and I do use it myself. There are features pytest has that make testing easier in some cases, but this project doesn't currently have a scenario that we **need** anything beyond what is already offered by Python's unittest module.

**Comment by kloczek** [comment]
> if your whole system depends on every python package you use to support your specific testing framework

Indeed .. issue is that I want to know before those areas of sensitivity before my client will come to me and will tell me "here it is something wrong" :)
In other words it is kind of proactive approach.
What will happen next is completely independent an probably would require some assessment about what to do with that.
Running few POCs I found some interesting results. What could be done next with those results is matter of running that framework on larger scale to be able identify some patterns and maybe design some other automation on top of that.

Going back to the ticket subject. Really please have look one more time on that case because it may be something very easy to fix for you  .. only this and nothing more :P
Please ..

**Comment by waylan** [comment]
We support and use `unittest` in the Python Standard Library. We will not be adding support for any other testing framework. The extra work would provide no benefit to the project itself as @facelessuser has explained multiple times. Please stop wasting our time with this.
